### PR TITLE
feat(storage): test for ribbon filter

### PR DESCRIPTION
**Pull request by Little-Wallace**
Signed-off-by: Little-Wallace <bupt2013211450@gmail.com>

I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

for https://github.com/risingwavelabs/risingwave/issues/7406

This PR is used to investigate the performance and false-positive rate of ribbon filter

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)
https://github.com/risingwavelabs/risingwave/issues/7406

### CODE REVIEW

**Comment by Little-Wallace** [comment]
A bad news is that ribbon filter seems worse than bloom-filter. If anyone is interested in this algorithm, I would be glad if you could help me run this benchmark in different OS and machine

**Comment by Little-Wallace** [comment]
image

I run it in my MBP M1-pro and I found that ribbon-filter could reach lower FPR but it would cost more time in query

File: src/storage/benches/bench_bloom_filter.rs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+fn bench_bloom_filter_read(c: &mut Criterion) {
+    let mut origin_data = Vec::with_capacity(TEST_COUNT);
+    let mut data = Vec::with_capacity(TEST_COUNT);
+    for idx in 0..TEST_COUNT {
+        let key = format!("test_000001_{:08}", idx);
+        data.push(Sstable::hash_for_bloom_filter(key.as_bytes()));
+        origin_data.push(key);
+    }
+    let mut rng = SmallRng::seed_from_u64(10244021u64);
+    let data = Bloom::build_from_key_hashes(&data[..(TEST_COUNT / 2)], 14);
```
Comment (KeXiangWang) [inline src/storage/benches/bench_bloom_filter.rs:18]: What does 14 here means?👀

Comment (Little-Wallace) [reply src/storage/benches/bench_bloom_filter.rs:18]: It is the default bits_per_key. I must make sure that the size of  the bits_per_key is same for both algorithm

Comment (KeXiangWang) [reply src/storage/benches/bench_bloom_filter.rs:18]: I'm not familiar with ribbon filter. But for bloom filter, bits_per_key is caculuated based on the given `number of keys` and `expected FPR`. `number of keys` is decided by the SST KV counts. `expected FPR` is a configurable parameter. It seems here the experiment is based on fixed `number of keys` and `bits_per_key`. Should we use `number of keys` and `expected FPR` instead?

**Comment by Little-Wallace** [comment]
A good news is that ribbon-filter would save 1/3 space cost than bloom-filter. For example, if bloom-filter generated 15MB  meta-data for 1GB data, ribbon-filter only need 10MB.

**Comment by soundOfDestiny** [comment]
> A good news is that ribbon-filter would save 1/3 space cost than bloom-filter. For example, if bloom-filter generated 15MB meta-data for 1GB data, ribbon-filter only need 10MB.

Should we tune parameters in order that ribbon-filter take the same cost as bloom-filter? seems it will then be a fair competition.

**Comment by Little-Wallace** [comment]
Ribbon-filter works best when the memory footprint is close. In this case bloom filter would cost 19 bits per key, ribbon-filter would not occur any false postive hit.

And when all algorithm want to reach the same FPR, 

```
bench_bloom_filter_read time:   [6.2200 µs 6.2678 µs 6.3241 µs]
fpr: 0.02316768%, cost: 608KB


bench_ribbon_filter_read time:   [8.1862 µs 8.2739 µs 8.3710 µs]
fpr: 0.00000000%, cost 617KB


bench_xor16_filter_read time:   [6.1159 µs 6.1970 µs 6.2946 µs]
0.00226281%, cost: 629KB
```

With the similar FPR, ribbon filter cost least memory but bloom-filter can not reach FPR of ribbon-filter or xor-filter no matter  how much I tune the value

```
bench_bloom_filter_read time:   [7.5708 µs 7.6208 µs 7.6717 µs]
fpr: 0.00538641%, cost: 1600KB

bench_ribbon_filter_read time:   [12.215 µs 12.341 µs 12.466 µs]
fpr: 0.00259066%, cost 488KB

bench_xor16_filter_read time:   [8.8213 µs 8.9364 µs 9.0462 µs]
fpr: 0.00225011%, cost: 629KB

**Comment by Little-Wallace** [comment]
Close because we has decide use xor-filter

**Pull request closed without merging**